[Evaluation][5] Add Evaluation entity with tests

dbczumar opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Add Evaluation entity with tests. Each Evaluation entity instance represents an evaluation of a set of inputs & outputs relative to some targets (optional). Evaluations include assessments from human / LLM judges, metrics (e.g. token usage), and tags (e.g. `"question_type": "SQL"`)

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

dbczumar committed:
Test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
test cases

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
smaller

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
test case

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
improved coverage

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
exclude

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add tag

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Entity and tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add eval

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
metric

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
merge

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Cases

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
merge

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix conflict

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove row

Signed-off-by: dbczumar <corey.zumar@databricks.com>

BenWilson2 commented on mlflow/entities/evaluation.py:
```diff
@@ -0,0 +1,192 @@
+        assessments: Optional[List[Assessment]] = None,
+        metrics: Optional[List[Metric]] = None,
+        tags: Optional[List[EvaluationTag]] = None,
+        error_code: Optional[str] = None,
```
nit: can we add doc strings for `error_code` and `error_message`?

dbczumar replied:
Absolutely - fixed!

BenWilson2 commented on mlflow/entities/evaluation.py:
```diff
@@ -0,0 +1,192 @@
+            return self.to_dictionary() == __o.to_dictionary()
+        return False
+
+    def to_dictionary(self) -> Dict[str, Any]:
+        """
+        Convert the Evaluation object to a dictionary.
+
+        Returns:
+            dict: The Evaluation object represented as a dictionary.
+        """
+        evaluation_dict = {
+            "evaluation_id": self.evaluation_id,
+            "run_id": self.run_id,
+            "inputs_id": self.inputs_id,
+            "inputs": self.inputs,
+        }
+        if self.outputs:
+            evaluation_dict["outputs"] = self.outputs
+        if self.request_id:
+            evaluation_dict["request_id"] = self.request_id
+        if self.targets:
+            evaluation_dict["targets"] = self.targets
+        if self.error_code:
+            evaluation_dict["error_code"] = self.error_code
+        if self.error_message:
+            evaluation_dict["error_message"] = self.error_message
+        if self.assessments:
+            evaluation_dict["assessments"] = [assess.to_dictionary() for assess in self.assessments]
+        if self.metrics:
+            evaluation_dict["metrics"] = [metric.to_dictionary() for metric in self.metrics]
+        if self.tags:
+            evaluation_dict["tags"] = [tag.to_dictionary() for tag in self.tags]
+        return evaluation_dict
```
WDYT about a comprehension to break up the if statements?

```python
def to_dictionary(self) -> Dict[str, Any]:
        """
        Convert the Evaluation object to a dictionary.
        Returns:
            dict: The Evaluation object represented as a dictionary.
        """
        evaluation_dict = {
            "evaluation_id": self.evaluation_id,
            "run_id": self.run_id,
            "inputs_id": self.inputs_id,
            "inputs": self.inputs,
            "outputs": self.outputs,
            "request_id": self.request_id,
            "targets": self.targets,
            "error_code": self.error_code,
            "error_message": self.error_message,
            "assessments": [assess.to_dictionary() for assess in self.assessments] if self.assessments else None,
            "metrics": [metric.to_dictionary() for metric in self.metrics] if self.metrics else None,
            "tags": [tag.to_dictionary() for tag in self.tags] if self.tags else None,
        }
        return {k: v for k, v in evaluation_dict.items() if v is not None}
```

dbczumar replied:
Great idea - applied this suggestion!

BenWilson2 commented on mlflow/entities/evaluation.py:
```diff
@@ -0,0 +1,192 @@
+        """
+        assessments = None
+        if "assessments" in evaluation_dict:
+            assessments = [
```
clean :) love it!

BenWilson2 commented on tests/entities/test_evaluation.py:
```diff
@@ -0,0 +1,185 @@
+    assert evaluation.tags == [tag]
+
+
+def test_evaluation_to_from_dictionary():
```
Do we need to ensure JSON serializability at all or is that inconsequential to these objs?

dbczumar replied:
Fortunately inconsequential. During serialization to artifacts, we stringify any non-JSON-serializable fields

BenWilson2 reviewed (approved):
small nits and a non-blocking question. LGTM once https://github.com/mlflow/mlflow/pull/12526/files#r1661295183 is addressed!

dbczumar committed:
Address comments

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented:
> small nits and a non-blocking question. LGTM once https://github.com/mlflow/mlflow/pull/12526/files#r1661295183 is addressed!

Thanks @BenWilson2! Addressed your comments and will merge when tests pass

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Lint

Signed-off-by: dbczumar <corey.zumar@databricks.com>
